main.go

```diff
@@ -372,5 +375,20 @@ func main() {
 		exitWithError(err, "Failed to reload HAProxy config")
 	}
 
+	if config.Envoy.UseGRPCAPI {
+		// TODO: Implement Sidecar graceful shutdown to allow draining the Envoy
```

Comment:
We don't ever want to tell Envoy we are shutting down. If we did that, we'd never be able to deploy Sidecar. We have to rely on Envoy continuing to service traffic with its current config until Sidecar comes back up. I know Envoy v1 has a timeout after which it will expire stuff if it doesn't hear about it, but we should just rely on that (assuming v2 is the same).

Reply:
No, the issue here is preventing a situation when Sidecar goes down immediately, even though it's in the middle of sending an update to Envoy via gRPC. While Envoy should, I hope, be smart enough to handle that as a transaction, I'm not sure. So, I could imagine Envoy ending up in an inconsistent state while Sidecar is busy starting up again.

By implementing graceful shutdown, the gRPC server will not allow new connections, but it will drain the in-flight ones before going down.

If you think this is a valid concern, I can try to implement it, just need to know what you think about the design.

Reply:
In theory a partial update is better than no update. This is all eventually consistent anyway and this was always the case on HAProxy so I am not really concerned about it.

Reply:
OK, sounds good! I removed the TODO.